[Quant] Add the operator of decomposed fake quant per channel

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121297

**Summary**
Add the operator of `quantized_decomposed.fake_quant_per_channel` and test the forward and backward of this op with comparing to ATen.

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_decomposed_fake_quant_per_channel
```

**Next Step**
Optimize the performance: from the generated code of forward and backward graph, the code didn't vectorize.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

leslie-fang-intel committed:
pass the correctness testing

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add the operator of decomposed fake quant per channel"


**Summary**
Add the operator of `quantized_decomposed.fake_quant_per_channel` and test the forward and backward of this op with comparing to ATen.

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_decomposed_fake_quant_per_channel
```

**Next Step**
Optimize the performance: from the generated code of forward and backward graph, the code didn't vectorize.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jerryzh168 commented:
thanks! can you a perf test locally and report the perf number comparison with the current torch fake quant op as well?

jerryzh168 reviewed (approved):
just saw the last comment, it's fine to test perf in next PR

jerryzh168 commented on test/inductor/test_cpu_repro.py:
```diff
@@ -1730,6 +1730,72 @@ def fn(test_args):
+            )
+            return res
+
+        def test_eager_aten_fake_quant(
```
nit: might be better to rename `test` to be more descriptive, e.g. `run`

leslie-fang-intel replied:
Thanks for the suggestion and changed the test name.

leslie-fang-intel committed:
Update on "[Quant] Add the operator of decomposed fake quant per channel"


**Summary**
Add the operator of `quantized_decomposed.fake_quant_per_channel` and test the forward and backward of this op with comparing to ATen.

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_decomposed_fake_quant_per_channel
```

**Next Step**
Optimize the performance: from the generated code of forward and backward graph, the code didn't vectorize.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
Here are the Inductor CPP backend generated code:

- Forward: https://gist.github.com/leslie-fang-intel/f6c4a8b3a98882acbfa465887539e9a0#file-fake_quant_per_channel_forward_generated_code-py
- Backward: https://gist.github.com/leslie-fang-intel/f6c4a8b3a98882acbfa465887539e9a0#file-fake_quant_per_channel_backward_generated_code-py

leslie-fang-intel commented:
> thanks! can you a perf test locally and report the perf number comparison with the current torch fake quant op as well?

Although current Inductor generated code didn't vectorize, I write a [benchmark](https://gist.github.com/leslie-fang-intel/4714d20e331e70f81da4de3211b9d7ce) to test its current performance and here are the results:

- ATen fq time is: 1.687239408493042
- decomposed qdq time is: 1.2523536682128906
- compiled decomposed qdq time is: 1.3800063133239746

Hi @jerryzh168, could you help take a look of the benchmark script and do you think this preliminary result looks good to you?

leslie-fang-intel commented:
@pytorchbot merge
